Bluetooth: Add Byteorder encode XX-bit macros

Thalley opened the pull request:
Add a series of macros to help encode little-endian values into an array. 
Modifies a couple of Bluetooth macros to use these new macros. 

The macros can easily be expanded to also handle big-endian and more bit variants (e.g. 54-bits)

Thalley committed:
include: Bluetooth: Add BT_BYTES_LIST_LEXX macros

Add a series of macros that can encode a XX-bit value
into array values. This can be used to easily encode
const array data.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Use BT_BYTES_LIST_LEXX macros for encoding of data

Modify the BT_UUID_XX_ENCODE and BT_LE_SUPP_FEAT_XX_ENCODE
macros to use the SYS_ENCODE_LEXX macros, instead of re-inventing
the encoding.

Also apply the macros for several other places.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

stephanosio commented:
Rebased to fix CI failure

dcpleung commented:
Is "encode" a bluetooth terminology? When I first looked at `SYS_ENCODE_LE16()`, I thought that it was to make something into a 16-bit number. Maybe we can use something like `SYS_LE16_TO_BYTE_ARRAY()`. `*_ENCODE_*` is not very descriptive unless you are already familiar with its function.

andyross reviewed (dismissed):
This is an API consisting of macros that take a word and expand as... a comma-separated list?  Not even an array literal or function invocation?  Is this really something we want to expose as an API to users?  Is anyone going to be able to read this and know what's going on?  Are there other systems with similar APIs?

I mean, down in a subsystem I guess it's more common to have DSL's with weird conventions, so it's not like this is hurting anything as a bluetooth API (though... AFAICT there's only one spot in our entire tree that uses it!).

But as an app API I think this gets a -1 from me unless there's a big chunk of code that needs it and can't use a more conventional alternative.  Isn't there a simpler way to get endianness conversion in odd byte sizes?

Thalley commented:
> Is "encode" a bluetooth terminology? When I first looked at `SYS_ENCODE_LE16()`, I thought that it was to make something into a 16-bit number. Maybe we can use something like `SYS_LE16_TO_BYTE_ARRAY()`. `*_ENCODE_*` is not very descriptive unless you are already familiar with its function.

`ENCODE` may be too generic in that sense. `TO_BYTE_ARRAY` is too specific, as it doesn't really create a byte array, but rather puts a value into a comma separated expansion. 

I'm open for any suggestions :)

Thalley commented:
> This is an API consisting of macros that take a word and expand as... a comma-separated list? Not even an array literal or function invocation? Is this really something we want to expose as an API to users? Is anyone going to be able to read this and know what's going on? Are there other systems with similar APIs?
> 
> I mean, down in a subsystem I guess it's more common to have DSL's with weird conventions, so it's not like this is hurting anything as a bluetooth API (though... AFAICT there's only one spot in our entire tree that uses it!).
> 
> But as an app API I think this gets a -1 from me unless there's a big chunk of code that needs it and can't use a more conventional alternative. Isn't there a simpler way to get endianness conversion in odd byte sizes?

That is a fair point. From I can tell, Bluetooth is the subsystem that uses this most. Found a big-endian usecase in the flash driver though: https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/flash/spi_flash_at45.c#L253-L255) and the LORA driver (https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/lora/sx126x.c#L153-L154), and one in the tests of the net sockets: https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/net/socket/af_packet/src/main.c#L403-L414

There are multiple other smaller pieces of code here and there that could also use this. 

In the end, there aren't going to be any large chunks for code that will be modified, nor any code optimizations, so I understand your reluctance. The same argument could of course be made for the `sys_put_` and `sys_get_` which also doesn't change anything large, nor provides any optimizations, and yet they were approved (and have become quite widely used afterwards).

dcpleung commented:
Maybe `SYS_LE16_TO_LE8_PARTS`? Though, probably need something after `SYS_` to act as a (semi-)namespace.

Thalley commented:
@andyross Are you keeping your opinion after seeing the other presented code pieces in a non-Bluetooth context that could use these macros?

Thalley commented:
@jhedberg any suggestion on how to move forward with this PR? Or should I just let it die?

jhedberg commented:
> @jhedberg any suggestion on how to move forward with this PR? Or should I just let it die?

Not really, unless @andyross has come to different thoughts

Thalley commented:
> I think the code becomes more readable with this PR. If the macros are not welcome as a public API, then we should keep them internally in the Bluetooth subsystem.
> 
> If you are asking for naming suggestions/hints; "**SPLAT_LE16**"? This is a nod to the Python/Ruby "splat" unpacking syntax. Otherwise: "**BYTES_LIST_LE16**"?: All C/C++ resources I look at use the term "list" for a comma-separated list of expressions. E.g. "An initializer is an initializer-list surrounded by braces." Or "**BYTES_CSV_LE16**"?

I'm definitely open for ideas for naming. The naming used so far in this PR is based on the `ENCODE` naming already used in BT, but something like `BYTES_LIST` also makes sense

Thalley commented:
> So... for the record I don't want to be the only one holding this up if the code needs to go in. My objections were aesthetic: I think macrobatics are fine, I think public APIs for shared code are fine, but this particular macrobatic trick is just too weird and special purpose to be worth putting in a public API. But if the sense of the project is otherwise, feel free to merge.
> 
> (No +1 from me though, heh. I personally think this belongs in the subsystem code that wants it.)

Thanks for your reply!
I would be fine with adding this to a public BT header, assuming we don't end up wanting in Zephyr later and then end up having duplicate functionality :) 

@jhedberg @alwa-nordic @jori-nordic Should I just move this to a BT header, and if so, which? It's currently used by `gap.h` and `uuid.h`

alwa-nordic commented:
> Should I just move this to a BT header, and if so, which? It's currently used by `gap.h` and `uuid.h`

A new header file, I think. I can't find any existing header that it would fit into. Can we have `zephyr/bluetooth/byteorder.h`?

Thalley commented:
> > Should I just move this to a BT header, and if so, which? It's currently used by `gap.h` and `uuid.h`
> 
> A new header file, I think. I can't find any existing header that it would fit into. Can we have `zephyr/bluetooth/byteorder.h`?

That's also my thought. `bluetooth.h` is of course quite generic in many ways, and it could fit in there. 

We could also consider a more generic header such as `bluetooth/util.h` if we don't mind duplicating the names of other public header files (which `byteorder.h` also would be).

Thalley commented:
Updated the PR to only introduce the macros in a Bluetooth header. Also added additional changes to where they are used to provide more evidence for their usefulness

Thalley commented:
Rebased to solve merge conflicts
